### DISCUSSION TITLE
Using a formatter for formatting (instead of lint rules)

### DISCUSSION THREAD

**Discussion by JoshuaKGoldberg** [DefinitelyTyped/DefinitelyTyped#65993] [_Misc]
**Update (October 9th, 2023)**: This work is done! As of #66235's merge on October 6th, all DefinitelyTyped's types files are formatted with dprint _(with a small number of temporary exceptions for build issues)_.

---

## Context

👋 Hi all! I've been working with the TypeScript team on overhauling DefinitelyTyped's usage of formatting and linting tools. Alongside the effort to finally migrate from TSLint to ESLint (https://typescript-eslint.io/linting/troubleshooting/tslint -> https://github.com/microsoft/DefinitelyTyped-tools/issues/648), we'd like to align with how much of the industry & many linter maintainers recommend using a formatter vs. a linter ([myself included](https://typescript-eslint.io/linting/troubleshooting/formatting "typescript-eslint article: What About Formatting?")).

This is a surfacing of a plan we've formulated that we think will get DefinitelyTyped onto using a proper formatter for formatting. I'd very much like to see the community's reaction: does this seem reasonable to you? What edge cases are we missing? Will this offend your sensibilities as a programmer?

## DefinitelyTyped's Formatting Context

DefinitelyTyped today only utilizes a few ESLint rules for code formatting. It does not use [Prettier](https://prettier.io), [dprint](https://dprint.dev), or other dedicated formatting tools.

We'd like to move DefinitelyTyped onto one of those tools for the following benefits:

- **Consistency**: so contributors don't have to adjust to different formatting preferences per types package.
- **Dev time**: formatters save contributor time from being spent on manually touching up files.
- **Diff sizes**: so developers stop accidentally (or intentionally) reformatting entire files in the course of making an unrelated change.
- **Easier TSLint to ESLint migration**: removing usage of formatting rules reduces the amount of logic we'll need to port from TSLint to ESLint

> ✋ **Please do not reply to this with formatting opinions.**
> We know using a formatter instead of a linter for formatting can be a contentious choice.
> [Bikeshedding](https://en.wikipedia.org/wiki/Law_of_triviality) on nuances of whitespace preferences is not useful.

## The Plan

1. Determine the [formatter choice](#formatter-choice) and [formatting configuration](#formatting-configuration)
2. Create and merge a DefinitelyTyped-tools pull request that removes all formatting rules
3. Simultaneously to (2), create and merge DefinitelyTyped pull requests that:
   - Runs the formatter on all files in DefinitelyTyped
   - ~Adds checking in formatting fixes to the merge bot~
   - Adds a GitHub Action that makes a new commit to format all files if needed after merges to `master`
4. Create and merge a DefinitelyTyped pull request that adds a [`.git-blame-ignore-revs` file](https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view) to ignore the large formatting commits

> Edit August 10th: @jakebailey, @sandersn, and I discussed the _"check in formatting fixes to the merge bot"_ bit in more detail. It's a complicated work item given GitHub's security concerns & approval flows, and dprint performance has been consistently excellent. We're going to table it until it's shown to be needed (i.e. conflicts from `master` temporarily being unformatted start arising frequently).

### Formatter Choice: dprint

The two most common standalone formatters in the JavaScript/TypeScript ecosystem today are:

- [dprint](https://dprint.dev): much faster, but less popular
- [Prettier](https://prettier.io): more popular, but slower

We plan on using **dprint** for formatting for its superior performance.

A quick comparison of running both tools and [pprettier](https://www.npmjs.com/package/pprettier) on DefinitelyTyped shows a 10 second vs. minutes-scale difference:

```plaintext
npx dprint fmt  141.52s user 9.29s system 1481% cpu 10.177 total
npx pprettier --write './types/**/*.{ts,mts,cts}'  8.97s user 3.18s system 21% cpu 56.725 total
npx prettier -w ./types  361.28s user 35.60s system 126% cpu 5:13.36 total
```

dprint is also able to use a cacheable file cache, which drops times down to under a second. We may look into using it in CI as a followup performance optimization.

Additionally, running dprint on as large a repository as DefinitelyTyped will provide it and SWC with helpful community dogfooding (e.g. https://github.com/swc-project/swc/issues/7187, https://github.com/dprint/dprint-plugin-typescript/issues/533). Note that dprint is essentially [`deno fmt`](https://deno.land/manual@v1.35.0/tools/formatter), so we're not worried about its ability to handle all of DefinitelyTyped.

### Formatting Configuration

DefinitelyTyped today already has a [`.prettierrc.json` Prettier configuration file](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/59041d42631e3d6691cfb961b1d161c4efe61c41/.prettierrc.json):

```jsonc
{
  "$schema": "http://json.schemastore.org/prettierrc",
  // Edit August 29th: Removed overrides. See below.
  // "arrowParens": "avoid",
  // "singleQuote": true,
  // "trailingComma": "onlyMultiLine",
  "tabWidth": 4,
  "printWidth": 120
}
```

Our plan is to use the closest equivalent [dprint TypeScript configuration](https://dprint.dev/plugins/typescript/config). This proposal does not include work to change those settings. No set of settings is perfect or will satisfy all users.

_Edit August 29th: Following https://github.com/DefinitelyTyped/DefinitelyTyped/pull/66238#discussion_r1279995917 & https://github.com/DefinitelyTyped/DefinitelyTyped/pull/66238#discussion_r1279998651, I removed the style overrides._ 

### Explainer: When Formatting Runs

Pull request CI builds will not be blocked on formatting, so as to not inconvenience contributors who don't run a formatter. However, ~two parts of the plan ensure the `master` branch stays fully formatted~ the `master` branch will be fully formatted most of the time:

- ~Most PRs are merged by the DT bot: which can apply apply formatting on each PR as part of its existing merge process.~
- ~For manual merges done by maintainers: the action that runs on merges to `master` will capture any formatting mishaps~ The action that runs after merges to `master` will quickly format it if needed.

> Edit August 10th: see earlier note on tabling mergebot changes.

As a follow-up, we may want to create a general-purpose PR bot to make it easier for contributors to apply formatting to a PR on their behalf. This is out of scope for the current proposal.

### Can Files Opt Out of Formatting?

Yes, but considering the advantages mentioned earlier in this proposal, we strongly advise against doing so. Individual files that are auto-generated can use [configuration file `excludes`](https://dprint.dev/plugins/typescript/#install-and-setup) and/or [`// dprint-ignore-file` file ignore comments](https://dprint.dev/plugins/typescript/#ignoring-files) to disable the formatter.

_Edit July 31st, 2023_: As of https://github.com/dprint/dprint/pull/719, dprint also supports nested configurations. Individual `types/` directories can reconfigure dprint if they absolutely need. We'll set this up for directories that are fully auto-generated.

### Is Formatting Performance Important?

Most users would not notice a difference in performance between dprint and Prettier, as most files aren't large enough (thousands of lines long) to cause a significant difference. However, the CI runs on the `master` branch should ideally be as fast as possible to minimize the chance of users branching off a `master` commit with incorrectly unformatted code.

### Tabs or Spaces?

One potential side effect of switching to a dedicated formatter is that individual projects that [prefer using tabs for accessibility](https://github.com/prettier/prettier/issues/7475) may find it harder to do so in the project. We'll seed the configuration file to disable formatting for existing packages that use tabs. Please let us know if that's not enough: we can try to brainstorm a workaround, such as adding the ability to have project-specific overrides to the formatter.

_Edit July 31st, 2023_: As of https://github.com/dprint/dprint/pull/719, dprint also supports nested configurations. Individual `types/` directories can reconfigure dprint if they prefer need. Our initial PR that converts DefinitelyTyped to dprint will use tabs in all packages that seem to prefer them.

Today, only 19 packages use tabs - as determined by finding all `types/` directories that have any file with a number of `\n\t`s greater than 1% of its total `\n` count. Therefore, this proposal is not changing how the vast majority (>99.9%) of types packages use spaces. Doing so is a potential large follow-up task we can look into.

<details>
   <summary>Script run to find types packages that use tabs</summary>
   
```js
import fs from 'node:fs/promises';
import path from 'node:path';

async function checkDirectoryPathForTabs(directoryPath) {
    const directoryItems = await readdirSafe(directoryPath);

    for (const directoryItem of directoryItems ?? []) {
        const directoryItemPath = path.join(directoryPath, directoryItem);
        if (!directoryItemPath.includes('.ts')) {
            continue;
        }

        const contents = await readFileSafe(directoryItemPath);
        if (!contents) {
            continue;
        }

        const newlineTabsCount = contents.match(/\n\t/g)?.length ?? 0;
        if (!newlineTabsCount) {
            continue;
        }

        const newlineCount = contents.match(/\n/g)?.length ?? 0;

        if (newlineTabsCount > newlineCount / 100) {
            console.log(directoryPath)
            return;
        }
    }
}

async function readdirSafe(dirPath) {
    try {
        return await fs.readdir(dirPath);
    } catch {
        return undefined;
    }
}

async function readFileSafe(filePath) {
    try {
        return (await fs.readFile(filePath)).toString();
    } catch {
        return undefined;
    }
}

for (const directory of await fs.readdir('./types')) {
    const directoryPath = path.join('./types', directory);

    await checkDirectoryPathForTabs(directoryPath)
}
```

</details>

<details>
   <summary>Package names logged by that script</summary>
   
```plaintext
types/anime-quotes-api
types/arcgis-js-api
types/bootbox
types/ceddl__ceddl-polyfill
types/classificator
types/dns-packet
types/do-not-zip
types/ej.web.all
types/gulp-help
types/gulp-inject
types/js2coffee
types/json-query
types/qbittorrent-api-v2
types/react-leaflet-markercluster
types/reefjs
types/steelseries
types/titanium
types/uuid4
types/vscode
```

</details>

### What About `.json` and `.md` Files?

> _(added July 31st 2023)_

This proposal doesn't intend to format those files with the formatter. We may look into that as a followup.

### What's the Difference Between a Formatter and a Linter?

A formatter quickly (re-)formats code without adjusting any code logic. A linter runs a set of discrete rules that each flag -and optionally suggest fixes- for a specific kind of code defect.

See https://github.com/readme/guides/formatters-linters-compilers for more context.

### Why split up the PRs so much?

Splitting up changes is easier on tooling:

* It's easier to review the diffs on github.com (the _Files_ view slows down at hundreds of files)
* Blocking CI failures for a package will only block that package's PR
    * Example difficulty: https://github.com/microsoft/DefinitelyTyped-tools/issues/719
* If we need to revert later on, we'll be able to do so more granularly

Our splitting strategy was:

1. Find the biggest packages (https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/65993#discussioncomment-6600387) and send their PRs early
2. Then, send a PR for each letter of the alphabet
    * This is arbitrary but makes it easier to associate PRs with their related packages

## Thanks

This proposal came out of a group thread with discussion from:

- [David Sherret](https://github.com/dsherret)
- [Jake Bailey](https://github.com/jakebailey)
- [John Reilly](https://github.com/johnnyreilly)
- [Josh Goldberg](https://github.com/JoshuaKGoldberg)
- [Nathan Shively-Sanders](https://github.com/sandersn)
- [Piotr Błażejewicz](https://github.com/peterblazejewicz)
- [Sebastian Silbermann](https://github.com/eps1lon)

Appreciation in particular to Jake Bailey and Nathan Shively-Sanders for advising from the TypeScript team, and David Sherret from dprint.

**Comment by bradzacher** [comment]
A few quick points re prettier v dprint from experience.

Dprint is nowhere near as strict as prettier is - it allows a lot of variations in formatting an expression statement.
Depending on what you're trying to enforce - this may or may not be a factor. Given you're migrating from formatting lint rules (which from my experience are even less strict) - I doubt that it matters too much.

In regards to DevX - it's highly unlikely contributors have the dprint extension installed - meaning they won't get format on save and will have to rely on CLI formatting instead. Not a massive deal but it is something to consider. OTOH a contributor is much more likely to have the prettier extension installed.

In regards to performance - a lot of people tout runtime performance as a big negative for prettier. However one big counter point to this is that if you're doing things correctly then you should only ever be running your formatter in `O(changeset)`, not `O(repo)`. For most changes this should make for a fraction of the scale and thus a fraction of a runtime - to the point that the two tools are the same.

For some context I've spent a lot of time working with prettier at scale (I did several prettier version bumps on Meta's >800k file megarepo) and perf wasn't ever an issue. It does suck the few times you need to do an `O(monorepo)` run - but these should be rare times.

---

That aside - really happy to see this.

Moving off of TSLint will be huge in of itself - and having formatter for the codebase will be great for keeping changes down (I've seen many people tweak indentation and destroy diffs!)

**Comment by JoshuaKGoldberg** [comment]
Status update (July 31st 2023): we've received only positive feedback on this proposal. So we're going to go for it. 🚀 

On the dprint side, @dsherret added nested configuration support _(👏 thank you!)_ in https://github.com/dprint/dprint/pull/719. That means we can specify using tabs in existing `types/` directories that prefer them (instead of disabling dprint in them altogether). I've updated the suggestion with an explicit _edit_ note.

My plan now is to:
1. Work on creating a PoC PR for this repository per step 3
    * Note that it might be split up into multiple PRs to help with reviewing
2. Once that PR is / those PRs are ready for review, creating a DefinitelyTyped-tools PR per step 2
3. Run all those PRs through review and broadcasting them on social media for visibility
4. Once those PRs are merged, continue to step 4

Note that we've only received explicit feedback from a couple dozen individuals. That's a relatively small fraction of even the active pool of DefinitelyTyped contributors. I'll be sure to include _"please give feedback even at this stage in the process"_ notes in all messaging to make it clear that _we do still want feedback_.

**Reply by jakebailey** [reply]
What projects are using tabs and why? My general feeling is that DT packages fit into two categories, "a human wrote it and put it on DT" and "this is generated / imported from another repo / build and should not be formatted". Is there actually anyone in the first camp that is intentionally using tabs? In which case, why? Is it arbitrary?

**Reply by JoshuaKGoldberg** [reply]
> 👋 Hi! If I tagged you in this, it's because your name is in the header for a package that uses tabs and/or we think is auto-generated.
> If you have the time, could you please let us know if you have any opinions or needs around formatting?
> At least, whether your package is auto-generated, and/or you were intentional about tabs vs. spaces?

| Package                       | Generated?    | Lines Scale | Spacing Mix   | Attributed To                                                                             |
| ----------------------------- | ------------- | ----------- | ------------- | ----------------------------------------------------------------------------------------- |
| `anime-quotes-api`            |               | 10s         | All Tabs      | _(removed in the file header)_                                                            |
| `arcgis-js-api`               |               | 10,000s     | Mixed[^1]     | @Esri, @andygup, @bsvensson                                                               |
| `bootbox`                     |               | 100s        | Mostly Spaces | @vbortone, @konpikwastaken, @kanup, @icereed, @trodi, @stannynuytkens, @renjfk            |
| `ceddl__ceddl-polyfill`       |               | 100s        | Mostly Spaces | @broekema41                                                                               |
| `classificator`               |               | 100s        | Mostly Spaces | @Nixinova                                                                                 |
| `dns-packet`                  |               | 100s        | Mostly Spaces | @jhurliman                                                                                |
| `do-not-zip`                  |               | 10s         | Mixed[^2]     | @alexandercerutti                                                                         |
| `ej.web.all`                  |               | 10,000s     | Mostly Spaces | @syncfusion                                                                               |
| `gulp-help`                   |               | 100s        | Mostly Spaces | @tkQubo                                                                                   |
| `gulp-inject`                 |               | 10s         | Mostly Spaces | @k-kagurazaka                                                                             |
| `js2coffee`                   |               | 100s        | All Tabs      | _(removed in the file header)_                                                            |
| `json-query`                  |               | 100s        | Mostly Spaces | @mtraynham                                                                                |
| `office-js`                   | Partially (?) | 100,000s    | Mostly Spaces | OfficeDev, @Rick-Kirkham, @AlexJerabek, @ElizabethSamuel-MSFT, @alison-mk, @samantharamon |
| `qbittorrent-api-v2`          |               | 100s        | Mixed[^2]     | @Totto16                                                                                  |
| `react-leaflet-markercluster` |               | 100s        | Mostly Spaces | @ciaccodavide                                                                             |
| `reefjs`                      |               | 100s        | Mostly Spaces | @shockdevv, @cferdinandi                                                                  |
| `steelseries`                 |               | 1,000s      | Mixed[^2]     | @ricky-rebo                                                                               |
| `titanium`                    | Partially (?) | 10,000s     | Mixed[^1]     | @appcelerator, @janvennemann, @drauggres, @ffMathy                                        |
| `uuid4`                       |               | 10s         | All Tabs      | @BrenekH                                                                                  |
| `vscode`                      | Yes           | 10,000s     | All Tabs      | "Visual Studio Code Team" (?), @microsoft                                                 |

[^1]: `index.d.ts` is tabs while tests are spaces.
[^2]: `index.d.ts` is spaces while tests are tabs.

**Reply by andygup** [reply]
Thanks @JoshuaKGoldberg, for `arcgis-js-api` we are looking at deprecating that package at the next release, so this might not be a problem any more for our package.

**Reply by andygup** [reply]
One more comment - we intentionally used tabs to keep the file size below 5MB.

**Reply by BrenekH** [reply]
`uuid4` can be formatted to spaces. I'm pretty sure it's tabs because that's just how my editor is configured, but I'd rather match the project as a whole.

**Reply by cferdinandi** [reply]
Tabs always, because then whoever is using it can decide how many spaces they want each tab to be. Anyone who disagrees is literally wrong. Fight me. 😂😂😂

**Reply by alexandercerutti** [reply]
Hi @JoshuaKGoldberg!
I think that having a shared common style is important as long as it is very well documented and controlled (e.g. tools to be used, rules honoring checks, and so on).

I wrote the definitions for `do-not-zip` (but that's not the only package - maybe another one or two, but I honestly do not remember which spacing type I used. I see they are mixed but I honestly don't know why).

I always found `tabs` to be more convenient for a matter of navigation, for a matter of file size (yeah, I know, it sounds like an excuse), because I rather see the arrows that show how much indented the code instead of the single dots in the editors (eh eh) and because you can adjust their space in your editor. Once I was using the conversion 1 Tab = 4 Spaces. Then I moved to 1 Tab = 2 Spaces and I didn't have to change my code and commit all the differences.

I always use tabs if I can choose (the same is valid for double quotes), especially in my personal projects. I usually fight to use tabs and double quotes if I must use the code of the project for a long time.

___

I do not create typings every day for Definitely Types, so I'm okay if you want to format everything to spaces, but I'd rather formatting be automatic when a PR is approved. I cannot have an extension for my editor for each system (I have Prettier installed and configured for both work and personal projects, I don't know if dprint has one).

**Reply by Totto16** [reply]
I just used the tools that were available at the time, it wasn't auto-generated, but run through a formatter (prettier I think, but I'm not sure) 
I personally prefer Tabs, but in a huge repo like this, it's best to use the standards, but If I'd have a vote, I'd vote for Tabs :bookmark_tabs: 
That said, if nearly everything in the repo uses spaces, I'm fine with it :+1:

**Reply by cferdinandi** [reply]
I'm also deeply confused about why I'm linked here and what this is even about. My project, ReefJS, uses tabs, but I don't  particularly care what anyone else uses in theres, nor will I be updating my project to meet anyone else's preferences.

**Reply by jakebailey** [reply]
@cferdinandi You are listed as a [maintainer](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/reefjs/index.d.ts#L4) of `@types/reefjs` and that also means you'd get pinged on PRs, discussions about it, etc, and this change would affect `@types/reefjs`.

If you aren't reviewing / maintaining these types, then we can drop the comment from the d.ts file to remove you from getting pinged; the types don't appear to have been updated in 2 years, it seems (and are still at v7.6, many many versions behind), which is probably why you hadn't noticed! 😄

**Reply by cferdinandi** [reply]
Yea, someone else definitely created that and submitted it on my behalf, because it wasn't me. Don't use Typescript!

**Reply by Rick-Kirkham** [reply]
@JoshuaKGoldberg We actually have 3 type definition folders in DefinitelyTyped. Besides the **office-js** that you noted, there's **office-js-preview** and **office-runtime**. 

You are correct that our *.d.ts files are partially autogenerated and partially manually written.

We'd like you to configure (or tell us how to configure) things so that dprint doesn't run at all in our 3 libraries. The primary reason is that, in addition to being a source Intellisense, our *d.ts files are one of multiple inputs to a complex system of generating reference documentation for our APIs. We don't want to risk that changes in the formatting of the *.d.ts files will break that system, or produce formatting in the documentation that is inconsistent with the formatting of other inputs into the system (such as a set of example code snippets that we maintain locally). 

In light of the nesting support in the latest dprint, I'll make a work item for us to install dprint locally and do some experiments with its configuration. If we find a pattern that works for us, we'll get back to you. But we are understaffed, so it may be many months before we have a chance to do that.

**Reply by JoshuaKGoldberg** [reply]
Great, thanks for the info @Rick-Kirkham! We'll add all three of those packages to the `"excludes"` section of `dprint.json`. They won't have formatting run on them at all.

**Reply by Rick-Kirkham** [reply]
@JoshuaKGoldberg We have one more that I didn't know about: custom-functions-runtime.

**Comment by JoshuaKGoldberg** [comment]
Here's a list of relevant issues and pull requests:

**DefinitelyTyped**:

* #66235
* #66522
* Letters of the alphabet:
  * #66238
  * #66523
  * #66524
  * #66525
  * #66526
  * #66527
  * #66528
  * #66529
  * #66530
  * #66531
  * #66532
  * #66533
  * #66534
  * #66535
  * #66536
  * #66537
  * #66538
  * #66539
  * #66540
  * #66541
  * #66542
  * #66543
  * #66544
  * #66545
  * #66546
  * #66552
* Individual or smaller groups of packages:
  * #66359
  * #66360
  * #66361
  * #66362
  * #66363
  * #66358
  * #66357
  * #66356
  * #66355
  * #66354
  * #66353
  * #66352
  * #66351
  * #66350
  * #66349
  * #66832 
  * #66834 
  * #66850
  * #66858

**DefinitelyTyped-tools**:

* https://github.com/microsoft/DefinitelyTyped-tools/issues/698
* https://github.com/microsoft/DefinitelyTyped-tools/pull/699
* https://github.com/microsoft/DefinitelyTyped-tools/pull/720
* https://github.com/microsoft/DefinitelyTyped-tools/pull/721
* https://github.com/microsoft/DefinitelyTyped-tools/pull/722
* https://github.com/microsoft/DefinitelyTyped-tools/pull/723
* https://github.com/microsoft/DefinitelyTyped-tools/pull/726
* https://github.com/microsoft/DefinitelyTyped-tools/pull/728

I'll keep it updated over time.

**Comment by JoshuaKGoldberg** [comment]
I did some shell scripting to find the biggest packages on DT. Running on macOS after removing all `node_modules` with `rm -rf node_modules types/*/node_modules types/*/*/node_modules` (+ a few more levels of `*/*` just in case)...

<details>
<summary>Directories &gt;1M</summary>

```shell
du -sh types/* | grep M | sort -nr
```

```plaintext
 25M    types/dojo
 17M    types/carbon__icons-react
 16M    types/node
 15M    types/openui5
 14M    types/devexpress-web
 13M    types/echarts
8.5M    types/fhir
8.2M    types/carbon__pictograms-react
6.4M    types/arcgis-js-api
5.2M    types/office-js-preview
5.1M    types/react-native
5.1M    types/activex-libreoffice
5.0M    types/office-js
4.9M    types/winrt-uwp
4.7M    types/lodash
4.1M    types/ignite-ui
3.4M    types/three
3.0M    types/ramda
2.9M    types/es-abstract
2.8M    types/ej.web.all
2.4M    types/extjs
2.3M    types/aws-sdk2-types
2.2M    types/sencha_touch
2.1M    types/jquery
2.0M    types/facebook-nodejs-business-sdk
1.9M    types/nginstack__engine
1.9M    types/docusign-esign
1.8M    types/hapi
1.8M    types/google-apps-script
1.7M    types/titanium
1.4M    types/react
1.4M    types/openfin
1.4M    types/activex-mshtml
1.3M    types/microsoft-graph
1.3M    types/lodash-es
1.3M    types/carbon-components-react
1.1M    types/winjs
1.1M    types/wechat-miniprogram
1.1M    types/p5
1.1M    types/openlayers
1.1M    types/fibjs
1.1M    types/consumer-data-standards
1.0M    types/selenium-webdriver
1.0M    types/crocks
```

</details>

<details>
<summary>Files &gt;1M</summary>

```shell
find types ~+ -type f | xargs du -sh | sort -nr
```

```plaintext
5.5M    types/arcgis-js-api/index.d.ts
5.2M    types/office-js-preview/index.d.ts
5.2M    types/dojo/dijit.d.ts
5.0M    types/activex-libreoffice/index.d.ts
4.9M    types/winrt-uwp/index.d.ts
4.9M    types/openui5/sap.m.d.ts
4.9M    types/office-js/index.d.ts
4.1M    types/ignite-ui/index.d.ts
3.0M    types/dojo/dojox.mobile.d.ts
2.9M    types/openui5/sap.ui.core.d.ts
2.4M    types/extjs/index.d.ts
2.2M    types/ej.web.all/index.d.ts
2.1M    types/sencha_touch/index.d.ts
1.9M    types/docusign-esign/index.d.ts
1.8M    types/fhir/r5.d.ts
1.7M    types/titanium/index.d.ts
1.6M    types/devexpress-web/v182/index.d.ts
1.6M    types/devexpress-web/v181/index.d.ts
1.5M    types/dojo/dojox.widget.d.ts
1.5M    types/devexpress-web/v172/index.d.ts
1.5M    types/devexpress-web/index.d.ts
1.4M    types/openui5/sap.ui.webc.main.d.ts
1.4M    types/dojo/index.d.ts
1.4M    types/dojo/dojox.form.d.ts
1.4M    types/devexpress-web/v201/index.d.ts
1.4M    types/devexpress-web/v192/index.d.ts
1.4M    types/devexpress-web/v191/index.d.ts
1.4M    types/devexpress-web/v171/index.d.ts
1.4M    types/activex-mshtml/index.d.ts
1.3M    types/microsoft-graph/index.d.ts
1.3M    types/fhir/r4b.d.ts
1.3M    types/fhir/r4.d.ts
1.2M    types/dojo/dojox.dgauges.d.ts
1.2M    types/devexpress-web/v162/index.d.ts
1.1M    types/dojo/dojox.mvc.d.ts
1.1M    types/dojo/dojox.grid.d.ts
1.1M    types/dojo/dojox.gauges.d.ts
1.1M    types/devexpress-web/v161/index.d.ts
1.0M    types/echarts/options/series/graph.d.ts
1.0M    types/dojo/dojox.calendar.d.ts
```

</details>

<details>
<summary><code>.md</code> files with the text <code>generate</code></summary>

```shell
grep  -e "generate" --include=\*.md -lnrw types
```

```plaintext
types/lodash/readme.md
types/simperium/README.md
types/meteor/globals/README.md
types/meteor/test/globals/README.md
types/meteor/README.md
types/material-ui/scripts/README.md
types/dojo/README.md
types/sawtooth-sdk/README.md
types/node/ts4.8/scripts/generate-inspector/README.md
types/node/ts4.8/scripts/generate-docs/README.md
types/node/scripts/generate-inspector/README.md
types/node/scripts/generate-docs/README.md
types/recharts/README.md
```

</details>

<details>
<summary>All the <code>types/*</code> folders referenced previously, as well as a couple of more that include a <code>scripts/</code> directory</summary>

```plaintext
types/activex-libreoffice
types/activex-mshtml
types/arcgis-js-api
types/aws-sdk2-types
types/carbon__icons-react
types/carbon__pictograms-react
types/carbon-components-react
types/consumer-data-standards
types/crocks
types/devexpress-web
types/docusign-esign
types/dojo
types/echarts
types/ej.web.all
types/es-abstract
types/extjs
types/facebook-nodejs-business-sdk
types/fhir
types/fibjs
types/get-intrinsic
types/google-apps-script
types/hapi
types/ignite-ui
types/jquery
types/locutus
types/lodash
types/lodash-es
types/material-ui
types/meteor
types/microsoft-graph
types/nginstack__engine
types/node
types/office-js
types/office-js-preview
types/oibackoff
types/openfin
types/openlayers
types/openui5
types/p5
types/ramda
types/react
types/react-native
types/recharts
types/sawtooth-sdk
types/selenium-webdriver
types/sencha_touch
types/simperium
types/three
types/titanium
types/wechat-miniprogram
types/winjs
types/winrt-uwp
```

At the very least, we can check if they're auto-generated, and exclude them from dprint if so.

**Reply by JoshuaKGoldberg** [reply]
I went through those lists of packages and this seems to be the list of auto-generated ones:

```plaintext
types/custom-functions-runtime
types/dojo
types/facebook-nodejs-business-sdk
types/fhir
types/fibjs
types/get-intrinsic
types/locutus
types/lodash
types/material-ui
types/meteor
types/nginstack__engine
types/node
types/office-js-preview
types/office-js
types/office-runtime
types/oibackoff
types/oibackoff
types/openui5
types/p5
types/sawtooth-sdk
types/three
types/winrt-uwp
types/vscode
```

For each potential types package, I looked at whether it...
* ...explicitly mentioned being auto-generated
* ...has a sibling generation script file
* ...mentions being auto-generated in a README.md file
* ...has been updated in the last year
* ...has recent commits/PRs that seem to be manually authored

_Edit:_ for the above seemingly auto-generated packages, these ones seem to have been updated recently (within the last year, by someone outside of the DT team):

```plaintext
custom-functions-runtime
facebook-nodejs-business-sdk
fhir
get-intrinsic
lodash
meteor
nginstack__engine
office-js-preview
office-js
office-runtime
openui5
p5
sawtooth-sdk
three
winrt-uwp
vscode
```

...while these were not updated recently by that definition:

```plaintext
dojo
fibjs
locutus
material-ui
oibackoff
sawtooth-sdk
winrt-uwp
```

---

The large packages that we aren't considering auto-generated _(until & unless we're told otherwise)_ are:

```plaintext
types/activex-libreoffice
types/activex-mshtml
types/arcgis-js-api
types/aws-sdk2-types
types/carbon__icons-react
types/carbon__pictograms-react
types/carbon-components-react
types/consumer-data-standards
types/crocks
types/devexpress-web
types/docusign-esign
types/echarts
types/ej.web.all
types/es-abstract
types/extjs
types/google-apps-script
types/hapi
types/ignite-ui
types/jquery
types/lodash-es
types/microsoft-graph
types/openfin
types/openlayers
types/ramda
types/react
types/react-native
types/selenium-webdriver
types/sencha_touch
types/titanium
types/wechat-miniprogram
types/winjs
```

_Edit:_ for the above seemingly manually authored packages, these ones seem to have been updated recently (within the last year, by someone outside of the DT team):

```plaintext
arcgis-js-api
aws-sdk2-types
carbon__icons-react
carbon__pictograms-react
consumer-data-standards
docusign-esign
ej.web.all
google-apps-script
lodash-es
microsoft-graph
ramda
react
react-native
selenium-webdriver
titanium
```

...while these were not updated recently by that definition:

```plaintext
activex-libreoffice
activex-mshtml
carbon-components-react
crocks
devexpress-web
echarts
es-abstract
extjs
hapi
ignite-ui
jquery
openfin
openlayers
sencha_touch
wechat-miniprogram
winjs
```

**Reply by bradzacher** [reply]
I'm not quite sure why it matters that it's aito-generated. They don't do any signed sourcs validation which means there's no reason that you can't run dprint over the code after generation so that it matches the codebase standard.

**Reply by JoshuaKGoldberg** [reply]
If we format for them on `master`, then every time they send a PR with updates, it'll be filled with irrelevant noise from formatting differences.

**Reply by bradzacher** [reply]
I'm not sure I understand - why couldn't they just run the formatter themselves afterwards?

1) generate types
2) run formatter
3) submit PR

If they don't want to do (2) then that's up to them to deal with the noise in the review, IMO.

**Reply by JoshuaKGoldberg** [reply]
They could, but we've been trying to minimize the impact of this change. At least one of the partner teams have explicitly indicated they don't have budget to update their side.

My bet is that we'll have 1-2 dozen entries in the `"excludes"`, then over the next few months we'll be able to trim it down to near-zero.

**Comment by JoshuaKGoldberg** [comment]
Now that folks have started talking about formatting preferences (https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/65993#discussioncomment-6597906 onward, https://github.com/DefinitelyTyped/DefinitelyTyped/pull/66238#discussion_r1279998651), here's a comparison of how different overrides in #66235's `dprint.json` file (corresponding to overrides in the `.prettierrc.json` impact the Git diff. Before all this, I ran `rm -rf types/*/node_modules; rm -rf types/**/node_modules`.

```plaintext
$ find types | grep .d.ts | wc -l
29104
$ npx dprint --version
dprint 0.40.1
$ git diff --shortstat
25100 files changed, 1829028 insertions(+), 1653017 deletions(-)
```

Before each alternative, I ran `git restore .`.

| Setting              | Baseline          | Alternative     | Files | Insertions | Deletions | Δ Files         | Δ Insertions       | Δ Deletions        |
| -------------------- | ----------------- | --------------- | ----- | ---------- | --------- | --------------- | ------------------ | ------------------ |
| `"*.useParentheses"` | `"preferNone"`    | `"force"`       | 25627 | 1835941    | 1659872   | 🔺 +527 (+2%)   | 🔺 +6,913 (+0.38%) | 🔺 +6,855 (+0.37%) |
| `"quoteStyle"`       | `"alwaysSingle"`  | `"alwaysDouble"` | 24805 | 1961398    | 1785401   | 🟢 -295 (-1%)   | 🔺 +132,370 (+7%)  | 🔺 +132,384 (+7%)  |
| `"trailingCommas"`   | `"onlyMultiLine"` | `"never"`       | 26754 | 1850403    | 1674797   | 🔺 +1,654 (-7%) | 🔺 +21,375 (-1%)   | 🔺 +21,780 (-1%)   |

> Fewer changes is considered better. 🟢 indicates fewer changes in an alternative. 🔺 indicates more changes in an alternative.  

@dsherret, @jakebailey, and @sandersn have all privately mentioned a desire to switch to double quotes. We could certainly do that. _edit: emphasis on the "could", as in this is a theoretical possibility_

**Reply by jakebailey** [reply]
Maybe I'm brainfarting, but I'm having trouble understanding this table. Is the green dot saying that choosing alwaysDouble will result in fewer changes?

**Reply by JoshuaKGoldberg** [reply]
Correct - and I've edited in a note saying that explicitly.

**Reply by JoshuaKGoldberg** [reply]
Following up: I removed the TypeScript-specific overrides from the `.dprint.jsonc` in #66235. We want to start anew with the defaults, and didn't see any strong motivation from any partner teams or existing code to stick with previous overrides.

**Comment by icholy** [comment]
This is a great idea. Some suggestions:

* use .git-blame-ignore-rev
* disable typescript-bot for these PRs.
